### ISSUE TITLE
How to properly deserialize dates?

### ISSUE THREAD

**Issue by ReneBeszon** [fastify/help#914]
#### You have already researched for similar issues?
Yes

#### What are you trying to achieve, or the steps to reproduce?

I'm trying to deserialize a properly typed and validated string attribute to a Date object.

Following definition resolves properly to "gueltigab" and "gueltigbis" being Dates.

```ts
import { FromSchema, JSONSchema } from "json-schema-to-ts";

export const NewCampaignSchema = {
    $id: "NewCampaign",
    type: "object",
    properties: {
        name: { type: "string" },
        description: { type: "string" },
        gueltigab: { type: "string", format: "date-time" },
        gueltigbis: { type: "string", format: "date-time" },
        templateparentId: { type: "number" }
    },
    additionalProperties: false,
    required: ["name", "gueltigab", "templateparentId"]
} as const satisfies JSONSchema;

export type NewCampaign = FromSchema<typeof NewCampaignSchema, { deserialize: [{ pattern: { type: "string", format: "date-time" }, output: Date }] }>;```

#### What was the result you received?

When using this schema inside a body schema validation of a route the validation works as expected, but the actual body doesn't contain a date object for the the attributes. Instead it still is the original string which leads to type errors later on.

#### What did you expect?

The attributes should be Date-objects.

#### Context

* *node version*: 18.13.0
* *fastify version*: 4.18.0

**Comment by metcoder95** [comment]
The `JSON Schema` spec specifies that your strings with `date-time` format should follow the [ISO8601 spec](https://json-schema.org/understanding-json-schema/reference/string.html#dates-and-times). Which is automatically returned if you transform the `new Date()` to string, or call `new Date().toISOString()`.

If you are looking for the whole `Date` object, you'll need to define it in that way (as an object), not as a format for a given `string` type

**Comment by ReneBeszon** [comment]
Hi Carlos,

unfortunately I don't understand your answer. From above schema and type the request handler correctly identifies that the attribute is a date (see screenshot below). But during execution the content is actually the string and subsequently fails, e.g. when saving to a database. On the other hand mapping is not possible as the typescript says it is Date and not a string.

```ts
interface RequestInterface {
    Body: BodyDataType,
    Querystring: QueryConfigSchemaType,
    Reply: ReplyDataType
};

const schema = {
    body: BodyDataSchema,
    querystring: QueryConfigSchema,
    response: { "200": ReplyDataSchema }
};

const handler: FastifyPluginAsync = async (fastify) => {

    fastify.post<RequestInterface>(`${baseUrl}/campaign`, { preValidation: [fastify.authenticate], schema: schema }, async (request, reply) => {

        logger.trace("Calling", { query: request.query, params: request.params, body: JSON.stringify(request.body) });

        await insertCampaigns(request.user.userId, request.user.tenantId, request.user.organisationId, request.user.userId, request.query.configId, request.body.data)
            .then((campaigns) => reply.send({ success: true, data: campaigns }))
            .catch((error: Error) => { logger.error("Error inserting campaign", { error: error }); reply.send({ success: false, message: error.message }) });
    });
}
```

![image](https://github.com/fastify/help/assets/70770935/70eead89-2895-4b74-a395-1a52573b7894)

**Comment by metcoder95** [comment]
Just to confirm that I understood you correctly. From the request, you expect a `date-time` string (that's clear). Then, after validation, you want to use its `Date` object representation.

Unless I'm missing something, you are forced to convert that `date-time` string into a `Date` object.
Something like:
```ts
const gueltigbis = new Date(request.body.gueltigbis)
```

I suggest you provide a 
[Minimum Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example) next time to be able to help you and understand your problem faster 🙂

**Comment by ReneBeszon** [comment]
Yes this is clearly my expectation as fastify already does it with the type information, please see my screenshot. There the type of the attribute "gueltigab" is Date and not string.

Luckily the Date-constructor supports a "Date" parameter and I'm able to apply this really ugly hack, but this is far from straight forward.

Please don't get me wrong, I do think that you have a very good framework, but it feels like your support for typescript just reaches about 80% which doesn't reflect your home page. Even when looking at plain javascript I would have thought this to be a feature, to be at least to configure such a serialization.

**Comment by mcollina** [comment]
> Please don't get me wrong, I do think that you have a very good framework, but it feels like your support for typescript just reaches about 80% which doesn't reflect your home page. Even when looking at plain javascript I would have thought this to be a feature, to be at least to configure such a serialization.

A few notes:
1. Ajv does not deserialize a string into a Date by default, this is done by another module, ajv-keywords, that you'd need to install separately and add to the fastify constructor option.
2. json-schema-to-ts _incorrectly_  assumes that conversion of strings to dates will happen.
3. Dates are not a JSON type, so I don't think it's correct to automatically deserialize them by default.

I think it's possible to configure everything to work, but this is not a typescript issue. I have seen a few people expecting dates to be automatically parsed as such a few times.

I think we should provide better docs for this, as it looks like a behvaior a few people would have.

**Comment by mcollina** [comment]
I'm sorry but I cannot reproduce. Consider the following:

```typescript
/// <reference types="@platformatic/service" />
import { FastifyInstance, FastifyPluginOptions } from 'fastify'
import { JsonSchemaToTsProvider } from '@fastify/type-provider-json-schema-to-ts'

export default async function (fastify: FastifyInstance, opts: FastifyPluginOptions) {
  const app = fastify.withTypeProvider<JsonSchemaToTsProvider>()
  app.post('/', {
    schema: {
      body: {
        type: 'object',
        properties: {
          name: { type: 'string' },
          birthdate: { type: 'string', format: 'date-time' }
        }
      } as const
    }
  }, async (request, reply) => {
    console.log(typeof request.body.birthdate, request.body.birthdate)
    // Prints string <ISO 8601 date>
    const birthdate : string = request.body.birthdate
    return { birthdate }
  })
}
```

Everything is correctly typed as a string. I don't see where there is the claim that this is not typed correctly.

**Comment by mcollina** [comment]
I have tried to configure Ajv to parse dates, and I couldn't (even in JS). JTD supports them, but not JSON Schema.
JSON Schema is a specification for JSON, so it makes absolute sense. 

If you want a validator that supports JS primitives with complete typing, use https://www.npmjs.com/package/fastify-type-provider-zod.

I don't think there is much to do here.
